sdlib/d/gc/slab.d

```diff
@@ -0,0 +1,74 @@
+module d.gc.slab;
+
+import d.gc.emap;
+import d.gc.extent;
+import d.gc.spec;
+
+enum InvalidBinID = 0xff;
+
+struct slabAllocGeometry {
+	Extent* e;
+	uint sc;
+	size_t size;
+	uint index;
+	void* address;
+
+	this(void* ptr, PageDescriptor pd, bool ptrIsStart) {
+		assert(pd.extent !is null, "Extent is null!");
+		assert(pd.isSlab(), "Expected a slab!");
+		assert(pd.extent.contains(ptr), "ptr not in slab!");
+
+		e = pd.extent;
+		sc = pd.sizeClass;
+
+		import d.gc.util;
+		auto offset = alignDownOffset(ptr, PageSize) + pd.index * PageSize;
+		index = binInfos[sc].computeIndex(offset);
+
+		auto base = ptr - offset;
+		size = binInfos[sc].itemSize;
+		address = base + index * size;
+
+		assert(!ptrIsStart || (ptr is address),
+		       "ptr does not point to start of slab alloc!");
```

Comment:
ptrIsStart is only used in an assertion, a strong sign it probably doesn't belong in there.